File: cpp/1047_Remove_All_Adjacent_duplicates_In_String.cpp
class Solution {
public:
	string removeDuplicates(string S) {
		stack<char> st;
		for (char c : S) {
			if (!st.empty() && c == st.top()) {
				st.pop();
			}
			else {
				st.push(c);
			}
		}
		string ans = "";
		while (!st.empty()) {
			ans += st.top();
			st.pop();
		}
		reverse(ans.begin(), ans.end());
		return ans;
	}
};

File: python3/1047-remove-all-adjacent-duplicates-in-string.py
class Solution:
    def removeDuplicates(self, s: str) -> str:
        stk = []
        for i in s:
            if stk:
                if stk[-1] == i:
                    stk.pop()
                else:
                    stk.append(i)
            else:
                stk.append(i)
        return ''.join(stk)

        # stk = []
        # for i in s:
        #     if stk and stk[-1] == i:
        #         stk.pop()
        #     else:
        #         stk.append(i)
        # return ''.join(stk)


s = 'abbaca'
print(Solution().removeDuplicates(s))
s = 'azxxzy'
print(Solution().removeDuplicates(s))
